In [6]:
a = 123
b =  123

print(a is b)
print(id(a))
print(id(b))

True
1648414752
1648414752


In [2]:
print(a ==b)

True


In [4]:
print(["사","바"] is ["사", "바"])
print(["사","바"] == ["사", "바"])


False
True


In [8]:
c = 12341234
d = 12341234
print(c is d)

False


In [9]:
a = -1
b = -1
print(a is b)

True


In [10]:
a= 1
b = (1)
c = (1,)
print(type(a))
print(type(b))
print(type(c))

<class 'int'>
<class 'int'>
<class 'tuple'>


In [13]:
def my_func(a, b):
    c = a + b
    print(c) 

result = my_func(1, 5)


print(result)

6
None


In [15]:
num = 3+4j 
num.imag

4.0

In [17]:
num = 1 +4j
num.real

1.0

- 시퀀스 자료형
순서대로 나열된 것이 정렬된 것은 아니다


In [18]:
result = 4 + True + False + 5

print(result)

10


In [19]:
word = "python"

indexing = word[3:8]

print(indexing)

hon


In [20]:
names = ['John', 'Ron', 'James', 'Betty']
print(names[-2][-2])

e


- L.append(0*5) 에서 0*5 는 0을 5번 복사하는 것이 아닌
단순히 0 곱하기 5연산을 실행해
L.append(0) 과 동일하게 L 리스트에 정수 0 하나 만을 추가 한다 

In [25]:
L = [] ; L.append(0*5)
print(L)

L = [0 for i in range(5)]
print(L)

L = [0] * 5
print(L)

[0]
[0, 0, 0, 0, 0]
[0, 0, 0, 0, 0]


In [26]:
s = "hello my name is ssafy"
for i in s:
    if i == "m":
        print(s)

hello my name is ssafy
hello my name is ssafy


In [27]:
string = "i'm hungry"
result = ""
for idx,value in enumerate(string):
    if idx%2:
        result += value.upper()
    else:
        result += value
print(result)

i'm hUnGrY


In [32]:
d1 = {'d': dict()}
d2 = dict(d={})
print(d1)
print(d2)
print(len(d1))
print(len(d2))

{'d': {}}
{'d': {}}
1
1


In [34]:
d = {'a': 1, 'b': 2}
a1 = d.update(c=3)
a2 = a1
print(a1)
print(a2)

None
None


In [36]:
fruits = { 'apple': '사과', 'banana': '바나나' }
a = fruits.get('apple')
b = fruits.get('cherry')
c = fruits.get('melon', True)
d = { a: b }
if c:
    print(d)
    
print(a)
print(b)
print(c)

{'사과': None}
사과
None
True


- 해설) 딕셔너리에서 get 메서드는 존재하는 key 값의 경우 value 를, 없는 key 면 에러가 아닌 None 을, 없는 key 와 두번째 인자를 넘기면, 해당 인자를 리턴한다.

- 함수의 인자(parameter)는 함수 선언시 설정한 값이며, 인수(argument)는 함수 호출시 넘겨주는 값이다.

- 함수는 오직 하나의 객체만 반환할 수 있다.


- 같은 이름의 함수를 선언하면, 함수 호출 시 나중에 선언된 함수가 실행된다.

- return을 작성하지 않으면 함수는 None 값을 자동으로 return하게 되며, 오류는 발생하지 않는다.

In [37]:
def func(a, b=1, c=2, *args, **kwargs):
    d = sum([n*2 for n in args if n > 2])
    e = sum([v*v for k, v in kwargs.items()])
    return a + b + c + d + e

print(func(9, 4, 2, 3, 1, 7, d=3, e=6))

80


In [42]:
def my_sum(a, b):
    c = a + b
    print(c)

result = my_sum(5, 8)
print(c)

13
True


In [39]:
print(result)

None


In [43]:
a = 1
def my_func_1():
    a = 5
    my_func_2()

def my_func_2():
    print(a, end='')

my_func_1()
print(a)


11


In [44]:
def func(c='5', *args):
    a, c, b = args
    return a + b + c

print(func('3','4','1','2'))

421


- 해설) 함수 func에 정의된 인자(parameters)는 넘겨 받은 인수(arguments)로 인하여, c='3', args=('4','1','2')가 된다. c의 기본값이 정의되어 있지만 인수가 넘어 왔으므로 '3'이 할당되고, *args는 가변인자이므로 몇 개가 오더라도 하나의 변수 args에 tuple로 할당된다. tuple에 할당된 값('4', '1', '2')가 변수 a, c, b에 순서대로 할당하게 되고, 이 과정에서 c에 할당된 '3'이 없어지고 '1'이 할당되게 된다. 따라서 a, b, c를 이어붙인 결과는 '421'이 된다.

In [45]:
def func(c, b, a):
    return a*b+c

print(func(2, 5, 4))

22


In [46]:
name = 'hong'
class Person:
     name = 'choi'
     def greeting(self):
          print(name)

p1 = Person()
p1.name = 'kim'
p1.greeting()

hong


In [48]:
name = 'hong'
class Person:
     name = 'choi'
     def greeting(self):
          print(self.name)

p1 = Person()
p1.name = 'kim'
p1.greeting()

kim


- 해설) isinstance를 활용하기 위해서는 인스턴스, 클래스 순으로 인자를 넘겨줘야한다. 즉, isinstance(my_int, int)로 확인 해야 한다. 또는 type(my_int) == int로 비교할 수 있다.

In [51]:
my_int=3
type(my_int)

int

In [53]:
class Person:
    population = 0
    
    def __init__(self, name):
        self.name = name
        Person.population += 1

p1 = Person('hong')
p2 = Person('kim')
p3 = Person('kang')

In [54]:
p3.population

3

In [55]:
p2.population

3

In [57]:
Person.population

3

In [58]:
import copy

list1 = [3, 'a', 'b']
list2 = [1, 2, list1]

list3 = list1[:]
list4 = copy.copy(list2)
list5 = copy.deepcopy(list2)

In [59]:
print(list3)

[3, 'a', 'b']


In [60]:
print(list4)

[1, 2, [3, 'a', 'b']]


In [61]:
print(list5)

[1, 2, [3, 'a', 'b']]


In [62]:
print(type({ 'a': 'apple' }))

<class 'dict'>


In [66]:
a = "python"
a[:]

'python'

In [67]:
a = "python"
a[::]

'python'

In [68]:
a = "python"
a[::3]

'ph'

In [70]:
a=[1,2,3]
a.pop()

3

In [71]:
print(a)

[1, 2]


In [72]:
a=[1,2,3]
a.index(2)

1

In [73]:
print(a)

[1, 2, 3]


In [75]:
a=[1,2,3]
b = a.sort()
print(b)

None


In [146]:
print(a)

[1, 2, 3]


In [147]:
a = [1,2,4,5,3]
a.sorted()

AttributeError: 'list' object has no attribute 'sorted'

In [79]:
a=[1,2,3]
b=a.reverse()
print(b)

None


In [80]:
print(a)

[3, 2, 1]


In [81]:
a=[1,2,3]
a.reversed()

AttributeError: 'list' object has no attribute 'reversed'

In [85]:
a=[1,2,3]
b = a.remove(2)
print(b)

None


In [84]:
print(a)

[1, 3]


In [86]:
d = dict(a=1, b=2)
print(d)

{'a': 1, 'b': 2}


In [88]:
d = {"a":3,"b":6,"c":9}
d.keys()

dict_keys(['a', 'b', 'c'])

In [89]:
d.values()

dict_values([3, 6, 9])

- 가변형 mutable

변경이 자유로운 형식

리스트, 딕셔너리

- 불변형 immutable

값을 초기화 하면 변경이 불가능

정수형, 실수형, 문자열, 튜플



#### 시퀀스 자료형

시퀀스는 데이터의 순서대로 나열된 형식을 나타낸다

순서대로 나열된 것이 정열되었다는 것은 아니다

기본적인 시퀀스 타입

리스트(list)

튜플(tuple)

레인지(range)

문자열(string)

바이너리(binary)

In [94]:
r = range(0, 10, 2)
print(r)
print(range(0,10,2))

range(0, 10, 2)
range(0, 10, 2)


In [97]:
num=4

print("홀") if num % 2 == 1 else print("짝")

짝


In [112]:
a = "hi my name"
a.capitalize()

'Hi my name'

In [113]:
print(a)

hi my name


In [114]:
a.title()

'Hi My Name'

In [115]:
print(a)

hi my name


In [116]:
a.upper()

'HI MY NAME'

In [117]:
print(a)

hi my name


In [119]:
"lollol".replace("o", "a")

'lallal'

In [121]:
"woooow".replace("o", "", 3)

'wow'

In [122]:
"apple".find("a")

0

In [132]:
"apple".find("p")

1

In [133]:
"apple".find("w")

-1

In [134]:
a = "my name is"
print(a.split())

['my', 'name', 'is']


In [135]:
a = "my name is"
print(a.split("m"))

['', 'y na', 'e is']


In [137]:
a=[1,2,3,4]
a.pop()


4

In [138]:
print(a)

[1, 2, 3]


In [144]:
a=[1,2,3,4]
a.pop(3)

4

In [145]:
print(a)

[1, 2, 3]


#### 딕셔너리 메소드 활용

#### 추가 및 삭제

- ,pop(key[ , default])

key가 딕셔너리에 있으면 제거하고 그 값을 돌려준다. 그렇지 않으면 default를 반환한다

default가 없는 상태에서 딕셔너리에 없으면 keyError가 발생한다



- .update()

값을 제공하는 key, value로 덮어 쓴다



- .get(key[ , default])

key를 통해 value를 가져온다

절대로 KeyError가 발생하지 않는다. default는 기본적으로 None이다



In [150]:
dic={"a":1,"b":2,"c":3}
dic.pop("b")

2

In [152]:
dic.pop("d")

KeyError: 'd'

In [155]:
dic.pop("d","ok")

'ok'

In [156]:
dic

{'a': 1, 'c': 3}

In [158]:
dic.get("a")

1

In [161]:
print(dic.get("d"))

None


In [163]:
dic.get("s",4)

4

In [164]:
store={
    "과자" :{
        "빼빼로":1500
    }
}

In [165]:
store["과자"]

{'빼빼로': 1500}

In [166]:
store["과자"]["빼빼로"]

1500

In [167]:
store={
    "과자": [
        "양파링", "홈런볼", "새우깡"
    ]

}

In [168]:
store["과자"]

['양파링', '홈런볼', '새우깡']

In [169]:
store["과자"]["양파링"]

TypeError: list indices must be integers or slices, not str

In [170]:
store["과자"][0]

'양파링'